docs/docs/Best_Practices/Message_Limits.md

```diff
@@ -0,0 +1,27 @@
+# Message Limits
+
+There are limits on how much data can be sent back and forth between the plugin and engine at a time. There are five scenarios where this comes into play:
+1. Inputs sent from plugin to engine, as arguments to a [Plugin Operation](/References/Plugin_Operations.md). For example, the schema-defined `Repository` object that is provided as input to plugin operations.
+2. Outputs sent back from the plugin to the engine, as the return values from plugin operations.
+3. Exception messages and call stacks thrown by plugin code. For example, the `message` field within [User Visible Errors](/Best_Practices/User_Visible_Errors.md).
+4. Inputs sent from plugin to engine, as arguments to a [Platform library](/References/Platform_Libraries.md) function. For example, the `message` field that is passed to `logger.debug`. 
+5. Outputs sent back from the plugin to the engine, as the return values from Platform Library functions. For example, the `stdout` resulting from a call to `libs.run_bash`. 
+
+For case 1 and 2, the total size of data must be less than 4 mebibytes (4 MiB).
+
+For case 3, the total size of data must be less than 128 kibibytes (128 KiB).
+
+For case 4 and 5, the total size of data must be less than 32 mebibytes (32 MiB).
+
+The actual size of this information at runtime is dependent on how the Python interpreter chooses to represent the information, so it's not always possible to know ahead of time what the exact size will be.
+
+Here are some examples of where problems may occur:
+1. Using `libs.run_bash` to print the entire contents of a large file to stdout.
+2. Using a single `logger` command with many pages of output.
+3. Throwing an exception with a large message or stack trace.
+
+## How to tell if the message size was exceeded
+When a plugin operation or platform library callback fails with a `RpcError`.
```

Comment:
Is there a more specific error message? `RpcError` is pretty generic

Reply:
Yeah, it is generic. Unfortunately, we can't be specific here without enabling all of the debugging that was previously enabled to uncover the problem. We discussed before and decided not to enable the logging because there was too much noise in the logs when enabling ALL level debugging for io.grpc.netty and io.netty.

For reference, without enabling ALL level debugging, we see this in the debug logs for the libs server message size:
```
[2021-07-06 11:50:47,151][DEBUG][io.netty.handler.codec.http2.Http2ConnectionHandler#processGoAwayWriteResult:919][Libs Server thread #1][] [id: 0x28e03081] Sending GOAWAY failed: lastStreamId '1', errorCode '2', debugData 'writeAddress(..) failed: Broken pipe'. Forcing shutdown of the connection.
io.netty.channel.StacklessClosedChannelException
 at io.netty.channel.AbstractChannel$AbstractUnsafe.write(Object, ChannelPromise)(Unknown Source)
Caused by: io.netty.channel.unix.Errors$NativeIoException: writeAddress(..) failed: Broken pipe
```

What if we mention the Http2ConnectionHandler exception instead of RpcError?

Reply:
I think `Http2ConnectionHandler` makes sense. What is shows in the `DelphixUserException` `output`?

Reply:
Actually, upon reproducing the limit error for the DUE, I found out that `Http2ConnectionHandler` is not logged unless DEBUG level logging is enabled. We currently have WARN level logging, so I'll leave this out of the docs, but add info on what the DUE looks like